gateway-messages/src/lib.rs

```diff
@@ -0,0 +1,149 @@
+pub struct Request {
+    pub version: u32,
+    pub request_id: u32,
+    pub kind: RequestKind,
+}
+
+#[derive(Debug, Clone, Copy, SerializedSize, Serialize, Deserialize)]
+pub enum RequestKind {
+    Ping,
+    // TODO do we want to be able to request IgnitionState for all targets in
+    // one message?
+    IgnitionState { target: u8 },
+    IgnitionCommand { target: u8, command: IgnitionCommand },
+}
+
+/// Messages from an SP to a gateway, specifically responding to a [Request].
+#[derive(Debug, Clone, Copy, SerializedSize, Serialize, Deserialize)]
+pub struct Response {
+    pub version: u32,
+    pub request_id: u32,
+    pub kind: ResponseKind,
+}
+
+// TODO: Not all SPs are capable of crafting all these response kinds, but the
+// way we're using hubpack requires everyone to allocate Response::MAX_SIZE. Is
+// that okay, or should we break this up more?
+#[derive(Debug, Clone, Copy, SerializedSize, Serialize, Deserialize)]
+pub enum ResponseKind {
+    Pong,
+    IgnitionState(IgnitionState),
+    IgnitionCommandAck,
+    Error(ResponseError),
+}
+
+#[derive(Debug, Clone, Copy, SerializedSize, Serialize, Deserialize)]
+pub enum ResponseError {
+    /// The [RequestKind] is not supported by the receiving SP; e.g., asking an
+    /// SP without an attached ignition controller for ignition state.
+    RequestUnsupported,
+}
+
+/* IGNORE - treat these as notes; will flesh this out in the near future
+
+/// Messages from an SP to a gateway, prompted by the SP itself; e.g., ignition
+/// state change or serial console output.
+#[derive(Debug, Clone, Copy, SerializedSize, Serialize, Deserialize)]
+pub struct SpMessage {
+    // TODO: expections on this field? E.g. for serial console, does it imply
+    // ordering, or can we live with "wait to send more console data until we
+    // get an ack" with just a single outstanding message?
+    pub msg_id: u32,
+    pub kind: SpMessageKind,
+}
+
+#[derive(Debug, Clone, Copy, SerializedSize, Serialize, Deserialize)]
+pub enum SpMessageKind {
+    // TODO: Is only sending the new state sufficient?
+    IgnitionChange { target: u8, new_state: IgnitionState },
+    SerialConsole(SerialConsole),
+}
+*/
+
+#[derive(Debug, Clone, Copy, SerializedSize, Serialize, Deserialize)]
+pub struct IgnitionState {
+    pub id: u16,
+    pub flags: IgnitionFlags,
+}
+
+bitflags! {
+    #[derive(SerializedSize, Serialize, Deserialize)]
+    pub struct IgnitionFlags: u8 {
+        // RFD 142, 5.2.4 status bits
+        const POWER = 0b0000_0001;
+        const CTRL_DETECT_0 = 0b0000_0010;
+        const CTRL_DETECT_1 = 0b0000_0100;
+        // const RESERVED_3 = 0b0000_1000;
+
+        // RFD 142, 5.2.3 fault signals
+        const FLT_A3 = 0b0001_0000;
+        const FLT_A2 = 0b0010_0000;
+        const FLT_ROT = 0b0100_0000;
+        const FLT_SP = 0b1000_0000;
+    }
+}
+
+#[derive(Debug, Clone, Copy, SerializedSize, Serialize, Deserialize)]
+pub enum IgnitionCommand {
+    PowerOn,
+    PowerOff,
+    // Nop
```

Comment:
Not that you think otherwise, but FWIW I don't anticipate MGS directing the SP to send NOP commands.